ENH: Adding `bit_count` (popcount)

ganesh-k13 opened the pull request:
### `bit_count` UFunc Implementation

Adding `np.bit_count`:
```py
>>> a = np.int32(1023).bit_count()
10
```

### ToDos:

- [x] Ask mailing list ([pipermail link](https://mail.python.org/pipermail/numpy-discussion/2021-July/081988.html))
- [x] Add docs in various places
- [x] Add release notes
- [x] Add annotations
- [x] Add UT
- [x] Handle signed ints
- [ ] Add benchmarks
- [x] Support windows builtins
- [x] ~Add an Object loop~
- [x] Add support for `np.int*.bit_count`
- [x] Windows 32 bit bug.

### Benchmarks:
<details>
<summary>bench_ufunc.UFunc.time_ufunc_types:</summary>

<pre> 
python3 runtests.py --bench bench_ufunc.UFunc.time_ufunc_types
Building, see build.log...                 
Build OK                                   
· Discovering benchmarks                   
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Benchmarking existing-py_usr_bin_python3
[  0.00%] ··· Importing benchmark suite produced output:
[  0.00%] ···· NumPy CPU features: SSE SSE2 SSE3 SSSE3* SSE41* POPCNT* SSE42* AVX* F16C* FMA3* AVX2* AVX512F? AVX512CD? AVX512_KNL? AVX512_KNM? AVX512_SKX? AVX512_CLX? AVX512_CNL? AVX512_ICL?
[ 50.00%] ··· Running (bench_ufunc.UFunc.time_ufunc_types--).
[100.00%] ··· bench_ufunc.UFunc.time_ufunc_types                                                                                                                                                ok
[100.00%] ··· =============== =============
                   ufunc                   
              --------------- ------------- 
                 bit_count      62.9±0.5μs 
              =============== =============
</pre>
</details>

### Future Enhancements
Add SIMD implementations for popcount. Apparently, it's efficient.

resolves #16325

eric-wieser commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -845,3 +845,24 @@ npy_rshift@u@@c@(npy_@u@@type@ a, npy_@u@@type@ b)
+#ifdef __GNUC__
+    return __builtin_popcount@c@(a);
+#else
+    a  = a - ((a >> 1) & 0x55555555);                 // add pairs of bits
+    a  = (a & 0x33333333) + ((a >> 2) & 0x33333333);  // quads
+    a  = (a + (a >> 4)) & 0x0F0F0F0F;                 // groups of 8
+    return (a * 0x01010101) >> 24;                    // horizontal sum of bytes
```
this looks correct only for when the integer type is 32-bits, unless I'm missing something.

ganesh-k13 replied:
Yep, you are right, I will change it by referring https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel

eric-wieser commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -4232,3 +4232,35 @@ def add_newdoc(place, name, doc):
 
     """)
+
+add_newdoc('numpy.core.umath', 'countbits',
+    """
+    Computes the number of 1-bits in ``x``.
+
+    Parameters
+    ----------
+    x : array_like, unsigned int
+        Input array.
+    $PARAMS
+
+    Returns
+    -------
+    y : ndarray
+        The corresponding number of 1-bits in the input.
+        $OUT_SCALAR_1
+
+    References
+    ----------
+    .. [1] https://stackoverflow.com/a/109025/5671364
+
+    .. [2] Wikipedia, "Hamming weight",
+           https://en.wikipedia.org/wiki/Hamming_weight
+
+    Examples
+    --------
+    >>> np.countbits(np.uint64(1023))
+    10
+    >>> np.countbits(np.arange(10, dtype=np.uint32))
+    array([0, 1, 1, 2, 1, 2, 2, 3, 1, 2], dtype=uint32)
+
+    """)
```
I think it's worth including `popcount` somewhere in the docstring, since that seems like something people are likely to be searching for.

ganesh-k13 replied:
Oh yeah, makes sense, will get indexed that way

BvB93 reviewed (commented):
``` diff
cosh: _UFunc_Nin1_Nout1[L['cosh'], L[8], None]
+ countbits: _UFunc_Nin1_Nout1[L['countbits'], L[5], None]
deg2rad: _UFunc_Nin1_Nout1[L['deg2rad'], L[5], None]
```
Can you add annotations for the new ufunc in `numpy/__init__.pyi`?
In case you're wondering, the three parameters are its `__name__`, `ntypes` and `identity`.

ganesh-k13 commented:
Thanks @BvB93 , was just going to ask where to ask annotations :)

mattip commented:
Could you link to the discussion where the function was proposed on the mailing list or in an issue? This PR feels a bit disconnected from the desire to avoid adding any more functionality to NumPy's namespace, especially functions that do not appear in the [array API spec](https://data-apis.org/array-api/latest/API_specification/index.html). Maybe this should live somewhere in SciPy?

mattip commented:
Ahh, I see there was a brief discussion in issue #16325. I will comment there.

eric-wieser commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -845,3 +851,30 @@ npy_rshift@u@@c@(npy_@u@@type@ a, npy_@u@@type@ b)
 /**end repeat**/
+
+
+/**begin repeat
+ *
+ * #type  = ubyte, ushort, uint, ulong, ulonglong#
+ * #STYPE = BYTE,  SHORT,  INT,  LONG,  LONGLONG#
+ * #c     = ,      ,       ,     ll,    ll#
+ */
+#undef MAGIC_ARRAY
+#if 0
+/**begin repeat1
+ * #len = 8, 16, 32, 64#
+ */
+#elif NPY_BITSOF_@STYPE@ == @len@
+    #define MAGIC_ARRAY B@len@
+/**end repeat1**/
+#endif
+
+NPY_INPLACE npy_@type@
+npy_@type@_countbits(npy_@type@ a)
+{
+    a = a - ((a >> 1) & MAGIC_ARRAY[0]);
+    a = ((a & MAGIC_ARRAY[1])) + ((a >> 2) & MAGIC_ARRAY[1]);
+    return (((a + (a >> 4)) & MAGIC_ARRAY[2]) * MAGIC_ARRAY[3]) >> (NPY_BITSOF_@STYPE@ - 1)*8;
+}
+
```
What happened to using `__builtin_popcount`? Here's how CPython does it:

https://github.com/python/cpython/blob/7569c0fe91dfcf562dee8c29798ecda74d738aa8/Include/internal/pycore_bitutils.h#L86-L130

ganesh-k13 replied:
Yeah, thanks for bringing it up. So there is GNU GCC and Clangs in-built popcount. Also there is MSVC. Is there any other popcount that is popular enough that we have to include?

ganesh-k13 replied:
I think we can take the SSE and AVX in a follow-up commit, they to have efficient popcounts

ganesh-k13 replied:
I have taken a note to add MSVC support

ganesh-k13 replied:
Added in 0ac9918(clang/gcc) and 8a9b6ef(Windows)

eric-wieser commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -55,6 +55,12 @@
+static const npy_uint8  B8[]  = {0x55,               0x33,               0x0F,               0x01};
+static const npy_uint16 B16[] = {0x5555,             0x3333,             0x0F0F,             0x0101};
+static const npy_uint32 B32[] = {0x55555555,         0x33333333,         0x0F0F0F0F,         0x01010101};
+static const npy_uint64 B64[] = {0x5555555555555555, 0x3333333333333333, 0x00F0F0F0FF0F0F0F, 0x0101010101010101};
```
This looks like a mistake to me:
```suggestion
static const npy_uint64 B64[] = {0x5555555555555555, 0x3333333333333333, 0x0F0F0F0F0F0F0F0F, 0x0101010101010101};
```

ganesh-k13 replied:
Oh sorry for missing this, will fix in same commit and some UT today

eric-wieser commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -4232,3 +4232,35 @@ def add_newdoc(place, name, doc):
+
+add_newdoc('numpy.core.umath', 'countbits',
+    """
+    Computes the number of 1-bits in ``x``.
```
```suggestion
    Computes the number of 1-bits in ``x``. Analogous to the builtin `int.bit_count`.
```

ganesh-k13 commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -845,3 +851,30 @@ npy_rshift@u@@c@(npy_@u@@type@ a, npy_@u@@type@ b)
+{
+    a = a - ((a >> 1) & MAGIC_ARRAY[0]);
+    a = ((a & MAGIC_ARRAY[1])) + ((a >> 2) & MAGIC_ARRAY[1]);
+    return (((a + (a >> 4)) & MAGIC_ARRAY[2]) * MAGIC_ARRAY[3]) >> (NPY_BITSOF_@STYPE@ - 1)*8;
```
Any suggestions on how to avoid a compiler warning on `(NPY_BITSOF_@STYPE@ - 1)*8`. I expected it to be computed at compile-time, but it actually throws a warning.

eric-wieser replied:
What's the warning?

Qiyu8 replied:
Perhaps you could refer to a fast C/C++ bit population count library called [libpopcnt ](https://github.com/kimwalisch/libpopcnt).

ganesh-k13 replied:
Thanks @Qiyu8 , the [library ](https://github.com/kimwalisch/libpopcnt/blob/master/libpopcnt.h)has a lot of useful stuff we can refer to.

ganesh-k13 replied:
> What's the warning?

@eric-wieser ,  https://github.com/numpy/numpy/pull/19355/checks?check_run_id=2926152038#step:4:4113,
we are getting a `warning: right shift count >= width of type`

ganesh-k13 replied:
I have fixed it in 975b8d1

BvB93 commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -4232,3 +4232,36 @@ def add_newdoc(place, name, doc):
+add_newdoc('numpy.core.umath', 'bit_count',
+    """
+    Computes the number of 1-bits in ``x``.
+    Analogous to the builtin `int.bit_count` or `popcount` in other languages.
```
```suggestion
    Analogous to the builtin `int.bit_count` or ``popcount`` in other languages.
```
Since popcount doesn't link to anything.
Also, a single concrete example of one of these "other languages" would be nice.

ganesh-k13 replied:
I have given C++ as an example as it has one in `std`, hope that's a good example.

eric-wieser commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -55,6 +55,12 @@
  */
 #include "npy_math_private.h"
 
+static const npy_uint8  S[]   = {1, 2, 4, 8, 16};
+static const npy_uint8  B8[]  = {0x55,               0x33,               0x0F,               0x01};
+static const npy_uint16 B16[] = {0x5555,             0x3333,             0x0F0F,             0x0101};
+static const npy_uint32 B32[] = {0x55555555,         0x33333333,         0x0F0F0F0F,         0x01010101};
+static const npy_uint64 B64[] = {0x5555555555555555, 0x3333333333333333, 0x0F0F0F0F0F0F0F0F, 0x0101010101010101};
```
What's `S` used for? Also, I think there's a case for adding a `npy_popcount(hh,h,,l,ll)` function to `npy_math.h`, and putting the tables there rather than here. In particular, we could then reuse the same function for implementing `bit_count` for `np.int32` scalars without having to go through ufunc overhead.

We use names of that form already here:

https://github.com/numpy/numpy/blob/47c5acf72571efe7cb731a51db594288e1dcc24c/numpy/core/src/npymath/npy_math_internal.h.src#L814

ganesh-k13 replied:
Oh I see, let me try moving it.

ganesh-k13 replied:
I have moved it in 975b8d1

BvB93 commented on numpy/core/tests/test_umath.py:
```diff
@@ -1840,6 +1840,24 @@ def test_reduction(self):
             btype = np.array([True], dtype=object)
             assert_(type(f.reduce(btype)) is bool, msg)
 
+    @pytest.mark.parametrize("input_dtype",
+        np.sctypes["int"] + np.sctypes["uint"])
```
```suggestion
    @pytest.mark.parametrize("input_dtype", np.typecodes["AllInteger"] + ["?"])
```
Suggestion (feel free to ignore): using `np.typecodes` allows you to use a 1-liner here.

Also, adding the bool dtype to the test-list would be a good idea.

ganesh-k13 replied:
Ohh nice, thanks @BvB93 , will add it.

ganesh-k13 replied:
Hey @BvB93 , just wanted to know if there is a neat way to get say `np.int32` from `i` or `np.int64` from `l`. I tried `map(np.dtype, np.typecodes["AllInteger"] + "?")`, but this gives dtype object not the integer class.
Not an issue actually, but wanted to know if there is a nice way to do it :)

BvB93 replied:
Sure, you can get the corresponding scalar type via, for example, `np.dtype("i").type`.

ganesh-k13 replied:
Ohh I just got it, I did `map(lambda c: np.dtype(c).type, np.typecodes["AllInteger"] + "?")`

ganesh-k13 replied:
refined in 81628e7

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -3537,6 +3537,7 @@ arcsinh: _UFunc_Nin1_Nout1[L['arcsinh'], L[8], None]
 arctan2: _UFunc_Nin2_Nout1[L['arctan2'], L[5], None]
 arctan: _UFunc_Nin1_Nout1[L['arctan'], L[8], None]
 arctanh: _UFunc_Nin1_Nout1[L['arctanh'], L[8], None]
+bit_count: _UFunc_Nin1_Nout1[L['bit_count'], L[5], None]
```
```suggestion
bit_count: _UFunc_Nin1_Nout1[L['bit_count'], L[10], None]
```
Small update to `ufunc.ntypes`, now that both signed and unsigned ints are supported.

ganesh-k13 replied:
Ohh right, I missed this, thanks.

ganesh-k13 commented:
The test failed in my machine also. We are hitting some edge case for a 16bit, 9 1s bit combo. https://github.com/numpy/numpy/pull/19355/checks?check_run_id=2954162115#step:4:3482 . Will try to figureout why

ganesh-k13 commented on numpy/core/src/umath/funcs.inc.src:
```diff
@@ -267,6 +267,38 @@ npy_ObjectClip(PyObject *arr, PyObject *min, PyObject *max) {
+
+        builtin_popcount_func = PyObject_GetAttrString(obj, "bit_count");
+        if (builtin_popcount_func != NULL) {
+            result = PyObject_CallFunction(builtin_popcount_func, "O", obj);
```
I get a strange:
```py
TypeError: int.bit_count() takes no arguments (1 given)
```

Will try to figure out why. Is the syntax incorrect?

ganesh-k13 replied:
Had to call with no args

ganesh-k13 commented:
I have added a native `numpy/core/_internal._bit_count`(to use in object loops for Python < 3.10) which is basically a Python version of:
https://github.com/python/cpython/pull/771/files#diff-1a6e70e2beeecad88840c67284ac4d54a36998029244771fcc820e801390726aR5332

Please let me know if this approach is ok

ganesh-k13 commented:
Hi all, 234edfe is purely guesswork :). I did not have a reference PR like LCM/GCD. So apologies if it's super wrong :). 
Summary of PR:
```py
# This works now:
>>> np.int32(1023).bit_count()
10
```

ganesh-k13 commented:
It seems we do not have a Python 3.10 build test in the pipelines. Will it make sense to add one in another PR?

ganesh-k13 commented on numpy/core/_methods.py:
```diff
@@ -51,6 +52,10 @@ def _prod(a, axis=None, dtype=None, out=None, keepdims=False,
 
+def _bit_count(a, out=None, where=True, casting='same_kind',
+          order='K', dtype=None, subok=True):
+    return umr_bit_count(a, dtype)
```
Missed sending the other args, will fix

ganesh-k13 commented:
Hi all, apart from a few fixes from my side, is there anything major remaining? A review of the object loop and `np.int*.bit_count` will be helpful as they seem they could use a bit of review. Also, do we hit the mailing list, etc for a wider opinion?

BvB93 commented on numpy/core/tests/test_umath.py:
```diff
@@ -1840,6 +1848,23 @@ def test_reduction(self):
 
+    @pytest.mark.parametrize("input_dtype_obj, bitsize",
+            zip(bitwise_types, bitwise_bits))
+    def test_popcount(self, input_dtype_obj, bitsize):
```
Adding a test for negative signed integers might be a good idea.
I believe those involve taking their absolute value, correct? Or do they raise?

ganesh-k13 replied:
We do an `abs()` yes. Will add it

seberg replied:
`np.iinfo(np.int64).min` (the lowest value) is a bit interesting, since `-int_min == int_min` (abs does nothing).  But I think that actually just works out here, because casting to uint will give the right result.

BvB93 commented on numpy/core/tests/test_umath.py:
```diff
@@ -1840,6 +1848,23 @@ def test_reduction(self):
+            msg = f"bit_count for {num}"
+            assert i == np.bit_count(input_dtype(num)), msg
+
+        a = np.array([2**i-1 for i in range(1, bitsize)], dtype = input_dtype)
```
```suggestion
        a = np.array([2**i-1 for i in range(1, bitsize)], dtype=input_dtype)
```
Style nit.

BvB93 commented on numpy/core/tests/test_umath.py:
```diff
@@ -1840,6 +1848,23 @@ def test_reduction(self):
+
+        a = np.array([2**i-1 for i in range(1, bitsize)], dtype = input_dtype)
+        bit_count_a = np.bit_count(a)
+        expected = np.arange(1, bitsize, dtype = input_dtype)
```
```suggestion
        expected = np.arange(1, bitsize, dtype=input_dtype)
```
More style nits.

BvB93 commented on numpy/core/_methods.py:
```diff
@@ -288,3 +289,8 @@ def _dump(self, file, protocol=2):
 def _dumps(self, protocol=2):
     return pickle.dumps(self, protocol=protocol)
+
+def _bit_count(a, out=None, *, where=True, casting='same_kind',
```
``` python
def bitcount(
    self,
    out: None | NDArray[Any] = ...,
    *,
    where: _ArrayLikeBool_co = ...,
    casting: _CastingKind = ...,
    order: _OrderKACF = ...,
    dtype: DTypeLike = ...,
    subok: bool = ...,
) -> NDArray[Any]: ...  # Or `Any` in the case of `np.generic`
```
Could you also add some basic annotations for `ndarray.bitcount` and `generic.bitcount`?

Adding dtype support is a bit more involved here, so I'd suggest to just copy/paste the annotations above and 
reserve further improvements for a (potential) future PR.

ganesh-k13 replied:
Sure @BvB93 , will add them.

ganesh-k13 replied:
Should the output be `-> integer | NDArray[Any]: ...` or does NDArray cover the 0-d?

BvB93 replied:
Eh, 0D arrays are a very tricky subject with the current lack of shape-typing (https://github.com/numpy/numpy/issues/19064) so I'd stick to just `NDArray[Any]`.

The problem with returning a union is that the effective signature of these objects is based on the intersection of all of its its members, _i.e._ it would make all operations that aren't supported by both `integer` and `ndarray` illegal (like iteration). 
All in all this would result in a plethora of false-positive errors for the end user.

The most practical way of dealing with this situation is by lying, _i.e._ ignoring the existence of 0D arrays. 
It's unfortunate, but also the lesser of (three or so) evils.

ganesh-k13 replied:
Hey @BvB93 , can we add it to numpy/typing/_ufunc.pyi under `_UFunc_Nin1_Nout1`?

BvB93 replied:
> Hey @BvB93 , can we add it to numpy/typing/_ufunc.pyi under `_UFunc_Nin1_Nout1`?

I'd just stick adding them as normal methods to `ndarray` and `generic`.

BvB93 commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -4232,3 +4232,36 @@ def add_newdoc(place, name, doc):
+
+    Examples
+    --------
+    >>> np.bit_count(np.uint64(1023))
```
```suggestion
    >>> np.bit_count(1023)
```
I suppose the explicit casting to an unsigned integer type is redundant now, no?

ganesh-k13 replied:
Oh yes, I missed this. Thanks

ganesh-k13 commented on numpy/typing/tests/data/reveal/ufuncs.py:
```diff
@@ -66,3 +67,14 @@
 reveal_type(np.matmul(AR_f8, AR_f8))  # E: Any
 reveal_type(np.matmul(AR_f8, AR_f8, axes=[(0, 1), (0, 1), (0, 1)]))  # E: Any
+
+reveal_type(np.bit_count.__name__)  # E: Literal['bit_count']
```
@BvB93 I hope this test is right, also would be happy to add other cases for `pass`, `fail`, etc if it makes sense for this case.

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -2854,6 +2854,17 @@ class ndarray(_ArrayOrScalarCommon, Generic[_ShapeType, _DType_co]):
+        order: _OrderKACF = ...,
+        dtype: DTypeLike = ...,
+        subok: bool = ...,
+    ) -> Any: ...
```
```suggestion
    ) -> NDArray[Any]: ...
```
It seems the output types of `ndarray.bit_count` and `generic.bit_count` got swapped around.

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -2998,6 +3009,17 @@ class generic(_ArrayOrScalarCommon):
+        order: _OrderKACF = ...,
+        dtype: DTypeLike = ...,
+        subok: bool = ...,
+    ) -> NDArray[Any]: ...
```
```suggestion
    ) -> Any: ...
```
See https://github.com/numpy/numpy/pull/19355/files#r677587591.

ganesh-k13 replied:
Fixed in same commit, thanks.

BvB93 commented:
After the latest fix in https://github.com/numpy/numpy/pull/19355/commits/6a89c8fc79401d31eec95fc18dc69321156c1200 I'd say that the annotations are in a good place.

ganesh-k13 commented:
Hey @seberg / @mattip , can you please have a look at this PR when free or in the Triage meeting.

mattip commented:
The [linter is unhappy](https://dev.azure.com/numpy/numpy/_build/results?buildId=18747&view=logs&jobId=a2bdd735-2cd0-5874-344f-2950fb45c949&j=a2bdd735-2cd0-5874-344f-2950fb45c949&t=43bcde9c-2634-50fc-b6ce-b6802de90ad4)

mattip commented:
I don't remember, did this hit the mailing list? It should be non-controversial since there is `int.bit_count` in Python.

ganesh-k13 commented:
Hey Matti, the linter is complaining this [line](cwre-85446c9997-s6jxg) and this [file](https://github.com/numpy/numpy/blob/1d6b6c94090d4d2a2e3fbff468b376cfabd59f3f/numpy/core/umath.py), is it ok if I refactor it a bit?

mattip commented on numpy/core/tests/test_umath.py:
```diff
@@ -1759,7 +1759,15 @@ def test_reduce(self):
 
-    bitwise_types = [np.dtype(c) for c in '?' + 'bBhHiIlLqQ' + 'O']
+    _all_ints_bits = [
+        np.dtype(c).itemsize * 8 for c in np.typecodes["AllInteger"]]
```
The linter complains about this line being one character too long which is fine.

ganesh-k13 replied:
Fixed in 15dd4fa

mattip commented on numpy/core/umath.py:
```diff
@@ -22,7 +22,7 @@
     'arccos', 'arccosh', 'arcsin', 'arcsinh', 'arctan', 'arctan2', 'arctanh',
     'bitwise_and', 'bitwise_or', 'bitwise_xor', 'cbrt', 'ceil', 'conj',
-    'conjugate', 'copysign', 'cos', 'cosh', 'deg2rad', 'degrees', 'divide',
+    'conjugate', 'copysign', 'cos', 'cosh', 'bit_count', 'deg2rad', 'degrees', 'divide',
```
Please reflow this to stay withing 80 characters

ganesh-k13 replied:
same as https://github.com/numpy/numpy/pull/19355#discussion_r681014549

mattip reviewed (commented):
Of the 4 things the linter found, I think only two should be changed. It is a shame the CI stops because of this. #19576 was merged to fix that, let's see what happens on the next CI run

charris commented:
@mattip Do we still need the gcc 4.8.8 test now that we have moved to manylinux2014?

seberg commented:
> And regarding the mailing list, should I start a thread?

Yes, that would be great.  E.g. I don't remember if there was some consensus around adding it as a method on `ndarray`.

seberg commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -4232,3 +4232,37 @@ def add_newdoc(place, name, doc):
+add_newdoc('numpy.core.umath', 'bit_count',
+    """
+    Computes the number of 1-bits in ``x``.
+    Analogous to the builtin `int.bit_count` or ``popcount`` in C++.
```
Should we mention that the sign is ignored (i.e. for signed integers the abs-value is used)?

ganesh-k13 replied:
I shall add it 👍

ganesh-k13 replied:
Fixed in 6d9acdd

seberg commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -845,3 +845,59 @@ npy_rshift@u@@c@(npy_@u@@type@ a, npy_@u@@type@ b)
+    return __builtin_popcount@c@(a);
+#elif defined(_MSC_VER) && NPY_BITSOF_@STYPE@ >= 16
+    return TO_BITS_LEN(__popcnt)(a);
+#endif
```
`#else` to not have dead code on MSVC?

ganesh-k13 replied:
Fixed in 15dd4fa

seberg commented on numpy/core/include/numpy/npy_math.h:
```diff
@@ -111,6 +111,18 @@ NPY_INLINE static float __npy_nzerof(void)
+static const npy_uint8  MAGIC8[]  = {0x55,               0x33,               0x0F,               0x01};
+static const npy_uint16 MAGIC16[] = {0x5555,             0x3333,             0x0F0F,             0x0101};
+static const npy_uint32 MAGIC32[] = {0x55555555,         0x33333333,         0x0F0F0F0F,         0x01010101};
+static const npy_uint64 MAGIC64[] = {0x5555555555555555, 0x3333333333333333, 0x0F0F0F0F0F0F0F0F, 0x0101010101010101};
```
These look like they should be in `npy_math_internal.h.src` with the function that uses them?

ganesh-k13 replied:
Fixed in 15dd4fa

seberg commented on numpy/core/src/umath/funcs.inc.src:
```diff
@@ -267,6 +267,38 @@ npy_ObjectClip(PyObject *arr, PyObject *min, PyObject *max) {
+    PyObject *result;
+
+    /* Try to use inbuilt popcount if available */
+    {
```
Do you prefer the additional scope, or is it just C89 compatibility (if the latter, might as well drop it)

ganesh-k13 replied:
I actually followed Erics GCD implementation. Sure, I shall remove it.

ganesh-k13 replied:
Fixed in 4e78f17

seberg commented on numpy/core/src/umath/funcs.inc.src:
```diff
@@ -267,6 +267,38 @@ npy_ObjectClip(PyObject *arr, PyObject *min, PyObject *max) {
+        PyErr_Clear();
+    }
+
+    /* Otherwise, use our internal one, written in python */
```
I don't like this fallback idea. `bit_count` is defined on Python integers, and as such it is pretty standardized already?  It will work without scalars (since we will implement `bit_count` for those.  I would think the default object fallback should thus be fine?

ganesh-k13 replied:
Oh I see. I thought of adding our implementation to support 3.8 and 3.9 as only 3.10 has the internal one and we can eventually remove it. Not sure if this is a standard, we can remove it yeah, what do you think is better?

seberg replied:
Hmm, that makes sense.  Although I am a bit worried we are adding unnecessary complexity that we can't easily remove later.

seberg replied:
I was not aware of gcd, it may open up weird corner cases like:
```
np.gcd(np.inf, np.pi, dtype=object)
```
which is an infinite loop.  But I am willing to be convinced that it is worthwhile for some reason.

ganesh-k13 replied:
Fixed in 4e78f17, I added a skip in tests as well and now we raise normal error

seberg commented on numpy/core/tests/test_umath.py:
```diff
@@ -1759,7 +1759,15 @@ def test_reduce(self):
+    bitwise_types = [
+        np.dtype(c) for c in '?' + np.typecodes["AllInteger"] + 'O']
+    bitwise_bits = [
+        2,  # boolean type
```
`2` feels a bit confusing for boolean, but I guess its fine.

ganesh-k13 replied:
This line `assert i == np.bit_count(input_dtype(num)), msg`  converts `input_dtype(<non zero>)` to `True`, hence bitcount stays `1`, so I added 2.

seberg commented on numpy/core/code_generators/generate_umath.py:
```diff
@@ -955,6 +955,13 @@ def english_upper(s):
+    Ufunc(1, 1, None,
+          docstrings.get('numpy.core.umath.bit_count'),
+          None,
+          TD(ints),
```
This means that the bitcount of `int8` will return `int8` and a bitcount of `uint8` will return `uint8`, etc.  I am wondering if anyone has opinions on this?  It seems reasonable, but right now I am unsure enough to pause.  In theory every bit count can be represented using `uint8` (which would be strange though?), but using the default integer for counting `uint8` bits always also seems a bit strange.  (Basically, I am just not sure that our prior art of homogeneous loops quite applies to bit counting).

seberg reviewed (commented):
A few questions (also about the API) mainly.  In general, I think this looks fine and personally I am happy with the addition.  The API details should hit the mailing list.

ganesh-k13 commented:
Thanks a lot for the review @seberg , I shall incorporate them and start a thread in the mailing list.

ganesh-k13 commented on numpy/core/code_generators/generate_umath.py:
```diff
@@ -955,6 +955,13 @@ def english_upper(s):
           TD(ints),
           TD('O', f='npy_ObjectLCM'),
           ),
+'bit_count' :
```
Fixed in 15dd4fa

ganesh-k13 commented on benchmarks/benchmarks/bench_ufunc.py:
```diff
@@ -4,7 +4,7 @@
 
 ufuncs = ['abs', 'absolute', 'add', 'arccos', 'arccosh', 'arcsin', 'arcsinh',
-          'arctan', 'arctan2', 'arctanh', 'bitwise_and', 'bitwise_not',
+          'arctan', 'arctan2', 'arctanh', 'bit_count', 'bitwise_and', 'bitwise_not',
```
I have skipped it for now as the full list will get refactored :(. `noqa` cannot be applied to full list with a single `noqa`

ganesh-k13 commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -795,3 +808,59 @@ npy_rshift@u@@c@(npy_@u@@type@ a, npy_@u@@type@ b)
+#if ((defined(__clang__) || defined(__GNUC__))) && NPY_BITSOF_@STYPE@ >= 32
+    return __builtin_popcount@c@(a);
+#elif defined(_MSC_VER) && NPY_BITSOF_@STYPE@ >= 16
+    return TO_BITS_LEN(__popcnt)(a);
```
It seems like:
> In 32-bit mode, there are no 64-bit general-purpose registers, so 64-bit popcnt isn't supported.

Hence the [failure](https://dev.azure.com/numpy/numpy/_build/results?buildId=18780&view=logs&j=c2d14800-8d4a-5342-9575-ba481005464b&t=834033e0-67ab-57a9-dad0-2389f178e6e7&l=878). I shall fix it.

ganesh-k13 commented:
Hey @seberg , thanks for the replies in the mail thread. The points that I could notice are as follows:
1. We can stick to implementing bit_count to just int*. so that would be something like `np.int32(1023).bit_count()`
2. Make the return value type `uint8`

Shall I refactor the code to reflect the same?

seberg commented:
I am not sure either of those is quite settled yet.  Not sure how to best settle it.  If you have an opinion, maybe suggest that to the mailing list, so anyone disagreeing/having serious doubts can raise their voice.

mattip commented:
There are a couple of linting errors

mattip commented:
I think the mailing list settled on adding this as an ndarray method and not as a top-level numpy api function.

ganesh-k13 commented:
Hi @mattip , is it the other way around, as there is:
> Should `np.ndarray.bit_count()` exist?  I tend against this; but we should have it on (integer) scalars

and:
>  I felt we could support it in the same lines NumPy Mathematical Functions, something like GCD perhaps, where we do not have `np.ndarray.gcd` but do have an `np.gcd`

mattip commented:
Sorry, I remembered wrong. My summary should have been "adding this as an int scalar method and not as a top-level numpy api function". 

If in the future there is a need, we can add it to the top-level `np` namespace, either `np.bit_count` or under some `np.bit` namespace. There was no support for adding it like `gcd`. There was no negative response either, but I interpret the lack of positive response as "no support".

ganesh-k13 commented:
Thanks @mattip that makes sense. Let me refactor the code to reflect only np.int*.bit_count.

ganesh-k13 committed:
ENH: Implementation of bit_count (popcount)

ganesh-k13 committed:
ENH: Add bit_count to integer scalar type

ganesh-k13 committed:
ENH: Annotations for bit_count

ganesh-k13 committed:
ENH, WIP: Documentation for bit_count

ganesh-k13 committed:
DOC: Added `bit_count` (#19355)

ganesh-k13 commented on numpy/core/_add_newdocs_scalars.py:
```diff
@@ -257,3 +257,19 @@ def add_newdoc_for_scalar_type(obj, fixed_aliases, doc):
         (-1, 4)
         """.format(ftype=float_name)))
+
+# XXX Neat way to get all names. Same as TODO above.
```
Is there a nice way to get all integer names?

ganesh-k13 committed:
BUG: Fixed windows 32 bit issue with no `__popcnt64`

ganesh-k13 commented:
Will add UT and bench, added it to TODO on description

ganesh-k13 commented on numpy/core/_add_newdocs_scalars.py:
```diff
@@ -257,3 +257,19 @@ def add_newdoc_for_scalar_type(obj, fixed_aliases, doc):
+# XXX Neat way to get all names. Same as TODO above.
+for int_name in ('ubyte', 'ushort', 'uint', 'ulonglong',
+                 'byte', 'short', 'int', 'long', 'longlong'):
+    add_newdoc('numpy.core.numerictypes', int_name, ('bit_count',
```
How is the build failing here? It's passing for me locally.
> AttributeError: type object 'int' has no attribute 'bit_count'

mattip replied:
I think `numpy.core.numerictypes.int` is the Python `int`, which is not a NumPy type. You probably don't want to write to its docstring.

BvB93 replied:
Don't forget about `intc` and `uintc` as well.

ganesh-k13 replied:
Oh nice, yeah @mattip ,
```py
>>> np.core.numerictypes.int
<class 'int'>
```

I guess it worked for me as I was using 3.10

Sure @BvB93 , will add it

ganesh-k13 replied:
I have added to `numpy.int32`, so `intc` also gets it.

ganesh-k13 commented:
Hi All, apologies for the delay. Will target the remaining items soon.

ganesh-k13 committed:
DOC: Refined docstring for bit_count

ganesh-k13 committed:
TST: Tests for bit_count

ganesh-k13 commented on numpy/core/tests/test_scalar_methods.py:
```diff
@@ -102,3 +102,20 @@ def test_roundtrip(self, ftype, frac_vals, exp_vals):
+    def test_bit_count(self):
+        for exp in [10, 17, 63]:
+            a = 2**exp
+            assert np.uint64(a).bit_count() == 1
```
We hit an overflow here. Will take a look.

ganesh-k13 replied:
Fixed with https://github.com/numpy/numpy/pull/19355/files#diff-e1d48225a79233f1670054e229f9b4c596929600773aca6d333a72da72897052R224

ganesh-k13 committed:
ENH, MAINT: Changed return type to uint_8 | Removed extra braces and fixed typo

ganesh-k13 commented:
Hi All, I have incorporated all changes discussed in the mailing list:
1. Adding this as an int scalar method.
2. Changed return type of all to `uint_8` as we do not have object loops now, hence return value cannot exceed 64

Please let me know if any other changes are needed. I will add benchmarks meanwhile.

ganesh-k13 commented on numpy/__init__.pyi:
```diff
@@ -3203,6 +3203,7 @@ class integer(number[_NBit1]):  # type: ignore
         __args: Union[L[0], Tuple[()], Tuple[L[0]]] = ...,
     ) -> int: ...
     def tolist(self) -> int: ...
+    def bit_count(self: _ScalarType) -> int
```
Oh, my bad, invalid syntax and did not run --mode=full. Will fix.

ganesh-k13 committed:
BUG: Fixed syntax of bit_count

ganesh-k13 commented:
Apologies for the delay again :(. The old builds are not there, so rebasing and triggering to try and repro the 32 bit and refguide issue.

ganesh-k13 committed:
DOC, BUG: Fixed bit_count example

ganesh-k13 committed:
DOC, BUG: (#19355) Removed bit_count from routines.math.rst | Improved release notes

ganesh-k13 committed:
BUG: Added type suffix to magic constants

ganesh-k13 committed:
ENH: Handle 32 bit windows popcount | Refactored popcount implementation to new function

ganesh-k13 commented:
Hi all, seems like all issues are solved. Thanks to @pearu for suggestions on the 32-bit windows overflow issue. 
Can you please review the changes?

ganesh-k13 committed:
MAINT: Refactor type_methods, separate integer definitions

ganesh-k13 commented:
@seberg / @mattip / @eric-wieser, can you please review the new changes when free.

seberg commented on doc/release/upcoming_changes/19355.new_feature.rst:
```diff
@@ -0,0 +1,13 @@
+`bit_count` to compute the number of 1-bits in an integer
```
Probably needs either double-ticks or a full name.

seberg commented on doc/release/upcoming_changes/19355.new_feature.rst:
```diff
@@ -0,0 +1,13 @@
+
+Computes the number of 1-bits in the absolute value of the input.
+This works on all the numpy integer types. Analogous to the builtin
+`int.bit_count` or `popcount` in C++.
```
Also doubleticks for both.

seberg reviewed (commented):
LGTM, small comments only: without the ufuncs, this really should be very straight forward.  I guess we don't really need the lowlevel math stuff for scalars, but it doesn't hurt.

The parallel code is unused now though?

Personally, I still would not mind adding the ufuncs, but probably better to discuss/consider as a followup.

BvB93 commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -2318,6 +2338,23 @@ static PyMethodDef @name@type_methods[] = {
+ */
+static PyMethodDef @name@type_methods[] = {
+    /* for typing; requires python >= 3.9 */
+    {"__class_getitem__",
+        (PyCFunction)numbertype_class_getitem,
+        METH_CLASS | METH_O, NULL},
```
The concrete `np.number` subclasses all override [numbertype_class_getitem_abc](https://github.com/numpy/numpy/blob/c92864091e5928d92bc109d1505febe35f3909f1/numpy/core/src/multiarray/scalartypes.c.src#L1809) with [numbertype_class_getitem](https://github.com/numpy/numpy/blob/c92864091e5928d92bc109d1505febe35f3909f1/numpy/core/src/multiarray/scalartypes.c.src#L1848), 
the latter always raising due to a lack of free subscription parameters.

ganesh-k13 committed:
DOC: Added double-ticks

ganesh-k13 commented:
The only thing I would consider missing is the benchmarks, I'll be happy to add it in this PR itself, but not sure where is the right place to add it. We can add it later if this is urgent as well, let me know your thoughts on this.

seberg commented:
Lets put this in, thanks @ganesh-k13.  Just in case anyone thought otherwise:  This PR does currently _not_ include ufunc versions of these functions.  Although it does lay some ground-work that would make creating those ufuncs much easier.
(I guess previous versions did include it.)

ganesh-k13 commented:
Thanks Sebastian. I edited the title to avoid confusion.

h-vetinari commented:
Hey everyone, I've started building 1.22.0rc1 for conda-forge, and apparently, the include-guards here don't fully work on windows:
```

As a comparison, I looked up the [guards](https://github.com/microsoft/STL/blob/3ba0477f3f3f6249c0f421af6852a4b6100fb549/stl/inc/limits#L1144-L1149) used in the microsoft STL.

h-vetinari commented:
Hey everyone, I've [started](https://github.com/conda-forge/numpy-feedstock/pull/248) building 1.22.0rc1 for conda-forge, and apparently, the include-guards here don't fully work on windows:
```
  numpy\core\src\npymath\npy_math_internal.h.src(894): error C3861: '__popcnt16': identifier not found
  numpy\core\src\npymath\npy_math_internal.h.src(890): error C3861: '__popcnt': identifier not found
  numpy\core\src\npymath\npy_math_internal.h.src(890): error C3861: '__popcnt': identifier not found
  numpy\core\src\npymath\npy_math_internal.h.src(900): error C3861: '__popcnt64': identifier not found
```

Do you want me to raise an issue for this?

PS. As a comparison, I looked up the [guards](https://github.com/microsoft/STL/blob/3ba0477f3f3f6249c0f421af6852a4b6100fb549/stl/inc/limits#L1144-L1149) used in the microsoft STL.

charris commented:
Curious. The wheels compile and run OK on azure Windows. I wonder what the difference is.

h-vetinari commented:
> Curious. The wheels compile and run OK on azure Windows. I wonder what the difference is.

On a hunch (and because this PR was merged just after #20234), I tested it again with vs2019 (conda-forge still uses vs2017 by default), and things pass then.

Though I'm kind of surprised that the CI in this PR was passing before #20234...? 🤔

charris commented:
We moved up to vs2019 on azure because  "vs2017 will will be deprecated Nov 15 and removed next Mar".  Azure was giving us pending deprecation notices. See #20234. That was 25 days ago, three days before this was merged.

h-vetinari commented:
> We moved up to vs2019 on azure because "vs2017 will will be deprecated Nov 15 and removed next Mar".

Yes, I'm aware. I'm not suggesting to revert that. It may just be that conda-forge needs to follow suit here - there's no point (IMO) trying to bridge a gap that MSFT is intentionally creating (and in their defense, they've been extremely conservative with the ABI, so there shouldn't be an argument not to move to vs2019 for basically anyone).

> I don't recall if it was passing before that.

Since the PR was not rebased, there are still CI results visible (cf. also the commits tab), and they all look green.

vscodegithubby commented:
Please, add 
countl_zero
countl_one
countr_zero
countr_one
has_single_bit

h-vetinari commented:
@vscodegithubby, please open a separate issue instead of commenting on >2yo merged PRs.

seberg commented:
If you want to request a new feature and open a new issue, make sure to have more arguments of why it is important enough to be in NumPy and maybe consider starting on the mailing list.

Without that, an issue will hardly be useful unless you know you have strong maintainer (or similar) support.  Since new features need champions who push them for a while _and_ be convincing to the community.
